design-proposals/klip-61-bootstrapped-table-subscriptions.md

```diff
@@ -0,0 +1,89 @@
+# KLIP 61 - Bootstrapped TABLE subscriptions
+
+**Author**: Natea Eshetu Beshada @nateab |
+**Release Target**: 0.27.0 |
+**Status**: _In Discussion_ |
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** We want users to be able to very easily with a single query get the current state from a `TABLE` and 
+simultaneously subscribe to any future changes to that `TABLE` by specifying `EMIT ALL`.
+
+## Motivation and background
+
+ksqlDB is a streaming database that allows users to work with data in motion (stream processing) and data at rest (databases).
+Thus, we have push queries which let you subscribe to a `TABLE` as it changes in real-time, and also pull 
+queries which retrieves the current state of the table, similar to a traditional database query. To further cement ksqlDB's
+positioning at the intersection of stream processing and traditional databases, it is necessary to develop features that 
+seamlessly combine the two. Adding bootstrapped `TABLE` subscriptions to ksqlDB achieves this purpose by allowing users to query their 
+data at rest up until now, while also subscribing to any later changes from data that is in motion. 
+
+Users can try to achieve this functionality by simply combining a `TABLE` pull query followed by a `TABLE` push query, but this 
+could result in gaps in the returned data since there are no guarantees that the push query will not miss any data after the 
+pull query is executed. Implementing this feature, we will keep track off the offsets under the hood for the user so that 
+we can guarantee there is no missing data. It will also be more clean and intuitive for users to have to only use one query
+to achieve this.
+
+## What is in scope
+
+- Queries should be supported on `TABLES` with `EMIT ALL` at the end.
+- Bootstrapped subscriptions should retrieve all current rows in the `TABLE` and subscribe to any further changes.
+- Bootstrapped subscriptions should not terminate until the user quits them.
+- Bootstrapped subscriptions will have feature parity with push queries v2, such as filters and projections.
+
+## What is not in scope
+
+- Bootstrapped subscriptions will not be supported on `STREAMS`. Instead, we will throw an informative error that tells the
+user to use `SET 'auto.offset.reset'='earliest';` with a push query in order to get all past data in the `STREAM` while 
+simultaneously subscribing to any future updates. This is because unlike `TABLES` which have a current state that a pull query retrieves,
```

Comment:
Maybe even "materialized current state".  It presumably also won't work for tables that are not materialized.

Reply:
good catch, added